creme/datasets/synth/concept_drift_stream.py

```diff
@@ -0,0 +1,128 @@
+
+import numpy as np
+
+from .. import base
+from ..synth import Agrawal
+from creme.utils.skmultiflow_utils import check_random_state
+
+
+class ConceptDriftStream(base.SyntheticDataset):
+    """Generates a stream with concept drift.
+
+    A stream generator that adds concept drift or change by joining two
+    streams. This is done by building a weighted combination of two pure
+    distributions that characterizes the target concepts before and after
+    the change.
+
+    The sigmoid function is an elegant and practical solution to define the
+    probability that each new instance of the stream belongs to the new
+    concept after the drift. The sigmoid function introduces a gradual, smooth
+    transition whose duration is controlled with two parameters:
+
+    - $p$, the position of the change.
+    - $w$, the width of the transition.
+
+    The sigmoid function at sample $t$ is $f(t) = 1/(1+e^{-4(t-p)/w})$.
+
+    Parameters
+    ----------
+    stream
+        Original stream
+    drift_stream
+        Drift stream
+    seed
+        If int, `seed` is used to seed the random number generator;
+        If RandomState instance, `seed` is the random number generator;
+        If None, the random number generator is the `RandomState` instance used
+        by `np.random`.
+    alpha
+        Angle of change used to estimate the width of concept drift change.
+        If set, it will override the width parameter. Valid values are in the
+        range (0.0, 90.0].
+    position
+        Central position of the concept drift change.
+    width
+        Width of concept drift change.
+
+    Notes
+    -----
+    An optional way to estimate the width of the transition $w$ is based on
+    the angle $\alpha$, $w = 1/ tan(\alpha)$. Since width corresponds to
+    the number of samples for the transition, the width is rounded to the
+    nearest smaller integer. Notice that larger values of $\alpha$ result in
+    smaller widths. For $\alpha > 45.0$, the width is smaller than 1 so values
+    are rounded to 1 to avoid division by zero errors.
+
+    Examples
+    --------
+    >>> from creme import synth
+
+    >>> dataset = synth.ConceptDriftStream(stream=synth.SEA(seed=42, variant=0),
+    ...                                    drift_stream=synth.SEA(seed=42, variant=1),
+    ...                                    seed=1, position=5, width=2)
+
+    >>> for x, y in dataset.take(10):
+    ...     print(x, y)
+    {0: 6.3942, 1: 0.2501, 2: 2.7502} False
+    {0: 2.2321, 1: 7.3647, 2: 6.7669} True
+    {0: 6.3942, 1: 0.2501, 2: 2.7502} False
+    {0: 8.9217, 1: 0.8693, 2: 4.2192} True
+    {0: 2.2321, 1: 7.3647, 2: 6.7669} True
+    {0: 8.9217, 1: 0.8693, 2: 4.2192} True
+    {0: 0.2979, 1: 2.1863, 2: 5.0535} False
+    {0: 0.2653, 1: 1.9883, 2: 6.4988} False
+    {0: 5.4494, 1: 2.2044, 2: 5.8926} False
+    {0: 8.0943, 1: 0.0649, 2: 8.0581} False
+    """
+
+    def __init__(self, stream: base.SyntheticDataset = Agrawal(seed=112),
```

Comment:
It's not good practice to set default values with mutable values. See [this](http://effbot.org/zone/default-values.htm) for an explanation why. The correct way to proceed is either to not specify a default value, which in my opinion makes sense in this case, or either to specify `None` and check `if is None` before setting the `Agrawal` default. Note that if you use the latter option you can reuse the same `seed` parameter you're passing to `ConceptDriftStream`.

@smastelini this applies to you too, I've seen you doing it with the trees during the last Zoom call :)

Reply:
Agree on the best practice for mutable attributes. However, I am missing its application in this case. Is the `Agrawal` generator mutable when the `seed` is set?

Reply:
Probably not! Just wanted to make you aware. Don't worry about it in this case :)

Reply:
> @smastelini this applies to you too, I've seen you doing it with the trees during the last Zoom call :)

Hi @MaxHalford, I do instantiate a regressor as the default attribute. In practice, I clone this model to replicate it to the leaves. Should I change it to `None`? No modifications are performed in the passed object. That's why I didn't bother much about it at the time.

Reply:
@smastelini it's a code smell to not use `None`. Users always find a way to get bitten by this :)